fvm/environment/contract_updater.go

```diff
@@ -437,26 +437,50 @@ func (updater *ContractUpdaterImpl) RemoveContract(
 	return nil
 }
 
-func (updater *ContractUpdaterImpl) Commit() ([]ContractUpdateKey, error) {
-	updatedKeys, updateList := updater.updates()
+func (updater *ContractUpdaterImpl) Commit() (ContractUpdates, error) {
+	updateList := updater.updates()
 	updater.Reset()
 
+	contractUpdates := ContractUpdates{
+		Updates:   make([]ContractUpdateKey, 0, len(updateList)),
+		Deploys:   make([]ContractUpdateKey, 0, len(updateList)),
+		Deletions: make([]ContractUpdateKey, 0, len(updateList)),
+	}
+
 	var err error
 	for _, v := range updateList {
-		if len(v.Code) > 0 {
+		var currentlyExists bool
+		currentlyExists, err = updater.accounts.ContractExists(v.Name, v.Address)
+		if err != nil {
+			return ContractUpdates{}, err
+		}
+		shouldDelete := len(v.Code) == 0
+
+		if shouldDelete {
+			// this is a removal
+			contractUpdates.Deletions = append(contractUpdates.Deletions, v.ContractUpdateKey)
+			err = updater.accounts.DeleteContract(v.Name, v.Address)
+			if err != nil {
+				return ContractUpdates{}, err
+			}
+		} else if currentlyExists {
```

Comment:
maybe combine the else if / else.  ContractExists don't need to be called when deleting.  SetContract can be call unconditionally